jwst/datamodels/tests/test_library.py

```diff
@@ -0,0 +1,141 @@
+    """
+    Helper function to modify the association at `example_asn_path`
+    by adding an attribute `attr` to the member list (at index
+    `member_index`) with value `value`. This is used to modify
+    the `group_id` or `exptype` of a certain member for some tests.
+    """
+    with open(example_asn_path, 'r') as f:
+        asn_data = load_asn(f)
+    asn_data['products'][0]['members'][member_index][attr] = value
+    with open(example_asn_path, 'w') as f:
+        json.dump(asn_data, f)
+
+
+def test_load_asn(example_library):
+    """
+    Test that __len__ returns the number of models/members loaded
+    from the association (and does not require opening the library)
+    """
+    assert len(example_library) == _N_MODELS
+
+
+@pytest.mark.parametrize("attr", ["group_names", "group_indices"])
+def test_group_with_no_datamodels_open(example_asn_path, attr, monkeypatch):
+    """
+    Test that the "grouping" methods do not call datamodels.open
+    """
+    # patch datamodels.open to always raise an exception
+    # this will serve as a smoke test to see if any of the attribute
+    # accesses (or instance creation) attempts to open models
+    def no_open(*args, **kwargs):
+        raise Exception()
+
+    monkeypatch.setattr(stdatamodels.jwst.datamodels, 'open', no_open)
+
+    # use example_asn_path here to make the instance after we've patched
+    # datamodels.open
+    library = ModelLibrary(example_asn_path)
+    getattr(library, attr)
+
+
+@pytest.mark.parametrize(
+    "asn_group_id, meta_group_id, expected_group_id", [
+        ('42', None, '42'),
+        (None, '42', '42'),
+        ('42', '26', '42'),
+    ])
+def test_group_id_override(example_asn_path, asn_group_id, meta_group_id, expected_group_id):
+    """
+    Test that overriding a models group_id via:
+        - the association member entry
+        - the model.meta.group_id
+    overwrites the automatically calculated group_id (with the asn taking precedence)
+    """
+    if asn_group_id:
+        _set_custom_member_attr(example_asn_path, 0, 'group_id', asn_group_id)
+    if meta_group_id:
+        model_filename = example_asn_path.parent / '0.fits'
+        with dm.open(model_filename) as model:
+            model.meta.group_id = meta_group_id
+            model.save(model_filename)
+    library = ModelLibrary(example_asn_path)
+    group_names = library.group_names
+    assert len(group_names) == 3
+    assert expected_group_id in group_names
+    with library:
+        model = library.borrow(0)
+        assert model.meta.group_id == expected_group_id
+        library.shelve(model, 0, modify=False)
+
+
+def test_asn_attributes_assignment(example_library):
+
+    # test that the association attributes are assigned to the models
+    with example_library:
+        for i in range(_N_MODELS):
+            model = example_library.borrow(i)
+            assert hasattr(model.meta.asn, 'pool_name')
```

Comment:
Would you update this test to define `pool_name` and `table_name` since by default any created `ImageModel` will pass these `hasattr` tests:
```python
>> import jwst.datamodels as dm
>> m = dm.ImageModel()
>> hasattr(m.meta.asn, 'table_name')
True
>> hasattr(m.meta.asn, 'pool_name')
True
```

Reply:
Thanks for pointing this out.  It revealed a bug in `_assign_member_to_model()` that I hope will resolve some of the other file naming issues I had been having